esp/platform/txsummary.cpp

```diff
@@ -65,100 +398,443 @@ bool CTxSummary::contains(const char* key) const
+{
+    if(!validateKey(key))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    set(key, new TxEntryStr(key, value, logLevel, group));
 
     return true;
 }
 
-CumulativeTimer* CTxSummary::queryTimer(const char* name)
+bool CTxSummary::set(const char* key, bool value, const LogLevel logLevel, const unsigned int group)
 {
-    if (!name || !*name)
-        return NULL;
+    if(!validateKey(key))
+        return false;
 
     CriticalBlock block(m_sync);
-    TimerValue& timer = m_timers[name];
 
-    if (!timer)
+    set(key, new TxEntryBool(key, value, logLevel, group));
+
+    return true;
+}
+
+bool CTxSummary::set(const char* key, int value, const LogLevel logLevel, const unsigned int group, const char* suffix)
+{
+    if(!validateKey(key))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    set(key, new TxEntryInt(key, value, logLevel, group));
+
+    return true;
+}
+
+bool CTxSummary::set(const char* key, __int64 value, const LogLevel logLevel, const unsigned int group, const char* suffix)
+{
+    if(!validateKey(key))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    set(key, new TxEntryInt64(key, value, logLevel, group, suffix));
+
+    return true;
+}
+
+bool CTxSummary::set(const char* key, unsigned int value, const LogLevel logLevel, const unsigned int group, const char* suffix)
+{
+    if(!validateKey(key))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    set(key, new TxEntryUInt(key, value, logLevel, group, suffix));
+
+    return true;
+}
+
+bool CTxSummary::set(const char* key, unsigned __int64 value, const LogLevel logLevel, const unsigned int group, const char* suffix)
+{
+    if(!validateKey(key))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    set(key, new TxEntryUInt64(key, value, logLevel, group, suffix));
+
+    return true;
+}
+
+bool CTxSummary::set(const char* key, double value, const LogLevel logLevel, const unsigned int group, const char* suffix)
+{
+    if(!validateKey(key))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    set(key, new TxEntryDouble(key, value, logLevel, group, suffix));
+
+    return true;
+}
+
+// Set a new entry named 'destKey' with it's value copied from the entry named 'sourceKey'
+// Group and LogLevel of the new entry are set as passed in.
+bool CTxSummary::setCopyValueOf(const char* destKey, const char* sourceKey, const LogLevel logLevel, const unsigned int group)
+{
+    if(!validateKey(destKey))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    if(!__contains(sourceKey))
+        return false;
+
+    TxEntryBase* sourceEntry = queryEntry(sourceKey);
+    Owned<TxEntryBase> newEntry = sourceEntry->cloneWithNewName(destKey);
+    newEntry->logLevel = logLevel;
+    newEntry->group = group;
+    set(destKey, newEntry.getLink());
+    return true;
+}
+
+
+CumulativeTimer* CTxSummary::queryTimer(const char* key)
+{
+    if(!validateKey(key))
```

Comment:
This is a change in behavior that could have bad consequences. It was safe, even if unwise, to assume that queryTimer would never return NULL as long as key was not empty. We now have multiple reasons for returning a potentially unhandled value. Perhaps it should throw an exception with an explanation instead.

Reply:
You think it is bad to change this to disallow duplicates and return null because existing code that calls this is not expecting a null result (even though it could have previously returned null under different circumstances)? Wouldn't it be undesirable for this to be the only function in the API that directly throws an exception (begging for consistency) since callers wouldn't be expecting that either?